clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -365,44 +430,90 @@ CompletableFuture<Void> commit(Map<TopicPartition, OffsetAndMetadata> offsets) {
     }
 
     @Override
     public void seek(TopicPartition partition, OffsetAndMetadata offsetAndMetadata) {
-        throw new KafkaException("method not implemented");
+        long offset = offsetAndMetadata.offset();
+        if (offset < 0) {
+            throw new IllegalArgumentException("seek offset must not be a negative number");
+        }
+
+        if (offsetAndMetadata.leaderEpoch().isPresent()) {
+            log.info("Seeking to offset {} for partition {} with epoch {}",
+                offset, partition, offsetAndMetadata.leaderEpoch().get());
+        } else {
+            log.info("Seeking to offset {} for partition {}", offset, partition);
+        }
+        Metadata.LeaderAndEpoch currentLeaderAndEpoch = this.metadata.currentLeader(partition);
+        SubscriptionState.FetchPosition newPosition = new SubscriptionState.FetchPosition(
+            offsetAndMetadata.offset(),
+            offsetAndMetadata.leaderEpoch(),
+            currentLeaderAndEpoch);
+        this.updateLastSeenEpochIfNewer(partition, offsetAndMetadata);
+        this.subscriptions.seekUnvalidated(partition, newPosition);
     }
 
     @Override
     public void seekToBeginning(Collection<TopicPartition> partitions) {
-        throw new KafkaException("method not implemented");
+        if (partitions == null)
+            throw new IllegalArgumentException("Partitions collection cannot be null");
+
+        Collection<TopicPartition> parts = partitions.size() == 0 ? this.subscriptions.assignedPartitions() : partitions;
+        subscriptions.requestOffsetReset(parts, OffsetResetStrategy.EARLIEST);
     }
 
     @Override
     public void seekToEnd(Collection<TopicPartition> partitions) {
-        throw new KafkaException("method not implemented");
+        if (partitions == null)
+            throw new IllegalArgumentException("Partitions collection cannot be null");
+
+        Collection<TopicPartition> parts = partitions.size() == 0 ? this.subscriptions.assignedPartitions() : partitions;
+        subscriptions.requestOffsetReset(parts, OffsetResetStrategy.LATEST);
     }
 
     @Override
     public long position(TopicPartition partition) {
-        throw new KafkaException("method not implemented");
+        return position(partition, Duration.ofMillis(defaultApiTimeoutMs));
     }
 
     @Override
     public long position(TopicPartition partition, Duration timeout) {
-        throw new KafkaException("method not implemented");
+        if (!this.subscriptions.isAssigned(partition))
+            throw new IllegalStateException("You can only check the position for partitions assigned to this consumer.");
+
+        Timer timer = time.timer(timeout);
+        do {
+            SubscriptionState.FetchPosition position = this.subscriptions.validPosition(partition);
+            if (position != null)
+                return position.offset;
+
+            updateFetchPositions(timer);
+        } while (timer.notExpired());
```

Comment:
Is it possible for us to do future.get(timer) to wait for the position to be updated? I think running a do-while loop just keep sending the event over and over again, which we probably don't want right?

Reply:
@lianetm, I think you're better equipped to answer this, too.